File: static/js/exporter.js
import { store } from "./state.js";
import { t } from "./i18n.js";
import { toast } from "./ui.js";

function triggerDownload(blob, filename) {
  const url = URL.createObjectURL(blob);
  const link = document.createElement("a");
  link.href = url;
  link.download = filename;
  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);
  setTimeout(() => URL.revokeObjectURL(url), 5000);
}

function hasContent() {
  return store.transcript.length > 0 || store.advice !== null;
}

function buildPayload() {
  return {
    exported_at: new Date().toISOString(),
    session_id: store.sid,
    metrics: store.metrics,
    transcript: store.transcript,
    advice: store.advice,
  };
}

function buildMarkdown() {
  const payload = buildPayload();
  const lines = [
    "# AI Sales Copilot — Session",
    "",
    `- **Exported:** ${payload.exported_at}`,
    `- **Session ID:** ${payload.session_id || "—"}`,
    `- **Messages:** ${payload.metrics.messages_total || 0} ` +
      `(client: ${payload.metrics.messages_client || 0}, ` +
      `manager: ${payload.metrics.messages_manager || 0})`,
    "",
    "## Transcript",
    "",
  ];
  if (!payload.transcript.length) {
    lines.push("_empty_");
  } else {
    for (const item of payload.transcript) {
      const who = item.role === "client" ? "Client" : "Manager";
      lines.push(`- **${who}:** ${item.text}`);
    }
  }
  lines.push("", "## Last AI advice", "");
  if (!payload.advice) {
    lines.push("_no advice yet_");
  } else {
    for (const key of Object.keys(payload.advice)) {
      lines.push(`- **${key}:** ${payload.advice[key] || "—"}`);
    }
  }
  return lines.join("\n");
}

function timestamp() {
  return new Date().toISOString().replace(/[:.]/g, "-");
}

export function exportJson() {
  if (!hasContent()) {
    toast(t("toast.exportEmpty"), "warn");
    return;
  }
  const blob = new Blob([JSON.stringify(buildPayload(), null, 2)], {
    type: "application/json;charset=utf-8",
  });
  triggerDownload(blob, `ai-copilot-session-${timestamp()}.json`);
  toast(t("toast.exportReady"), "success");
}

export function exportMarkdown() {
  if (!hasContent()) {
    toast(t("toast.exportEmpty"), "warn");
    return;
  }
  const blob = new Blob([buildMarkdown()], { type: "text/markdown;charset=utf-8" });
  triggerDownload(blob, `ai-copilot-session-${timestamp()}.md`);
  toast(t("toast.exportReady"), "success");
}

File: services/exporter.py
from __future__ import annotations

import json
from datetime import datetime, timezone

from services.state_manager import SessionState

def _fmt_ts(epoch: float) -> str:
    return datetime.fromtimestamp(epoch, tz=timezone.utc).strftime("%Y-%m-%d %H:%M:%S UTC")

def session_to_dict(session: SessionState) -> dict:
    return {
        "session_id": session.sid,
        "started_at": _fmt_ts(session.started_at),
        "profile": session.client_profile.to_dict(),
        "metrics": session.metrics().to_dict(),
        "transcript": [m.to_dict() for m in session.messages],
        "last_advice": session.last_advice or {},
    }

def session_to_json(session: SessionState) -> str:
    return json.dumps(session_to_dict(session), ensure_ascii=False, indent=2)

def session_to_markdown(session: SessionState) -> str:
    metrics = session.metrics()
    duration_sec = int(metrics.duration_sec)
    duration_str = f"{duration_sec // 60:02d}:{duration_sec % 60:02d}"

    lines: list[str] = [
        "# AI Sales Copilot — Session export",
        "",
        f"- **Session ID:** `{session.sid}`",
        f"- **Started:** {_fmt_ts(session.started_at)}",
        f"- **Duration:** {duration_str}",
        f"- **Messages:** {metrics.messages_total} "
        f"(client: {metrics.messages_client}, manager: {metrics.messages_manager})",
        "",
        "## Client profile",
        "",
    ]

    profile = session.client_profile.to_dict()
    if any(profile.values()):
        for key, value in profile.items():
            lines.append(f"- **{key}:** {value or '—'}")
    else:
        lines.append("_empty_")
    lines.append("")

    lines.append("## Transcript")
    lines.append("")
    if not session.messages:
        lines.append("_no messages yet_")
    else:
        for msg in session.messages:
            who = "Client" if msg.role == "client" else "Manager"
            stamp = datetime.fromtimestamp(msg.timestamp, tz=timezone.utc).strftime("%H:%M:%S")
            lines.append(f"- `{stamp}` **{who}:** {msg.text}")
    lines.append("")

    lines.append("## Last AI advice")
    lines.append("")
    advice = session.last_advice or {}
    if not advice:
        lines.append("_not generated yet_")
    else:
        for key, value in advice.items():
            lines.append(f"- **{key}:** {value or '—'}")

    lines.append("")
    return "\n".join(lines)
